src/components/FileUploader.tsx

```diff
@@ -0,0 +1,384 @@
+    align-self: end;
+    margin-right: 0;
+    margin-left: auto;
+  }
+`;
+
+const { Dragger } = Upload;
+const { confirm } = Modal;
+
+enum STATUS {
+  WAITING,
+  READY,
+  STARTED,
+  PAUSED,
+}
+
+type Props = {
+  assetId?: number;
+  validExtensions?: string[];
+  onUploadSuccess: (file: UploadFileMetadataResponse) => void;
+  onUploadFailure: (error: string) => void;
+  onCancel: () => void;
+  beforeUploadStart: () => void;
+};
+
+type State = {
+  uploadStatus: STATUS;
+  fileList: any[];
+};
+
+const defaultState = {
+  uploadStatus: STATUS.WAITING,
+  fileList: [],
+};
+
+class FileUploader extends React.Component<Props, State> {
+  public static defaultProps = {
+    onUploadSuccess: () => {},
+    onUploadFailure: message.error,
+    onCancel: () => {},
+    beforeUploadStart: () => {},
+  };
+
+  currentUpload: any;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      uploadStatus: STATUS.WAITING,
+      fileList: [],
+    };
+    this.currentUpload = null;
+  }
+
+  getMIMEType = (fileURI: string) => mime.lookup(fileURI);
+
+  setupFilesBeforeUpload = (file: any) => {
+    if (
+      this.props.validExtensions === undefined ||
+      this.props.validExtensions.length === 0 ||
+      this.props.validExtensions.includes(
+        file.name
+          .split('.')
+          .pop()
+          .toLowerCase()
+      )
+    ) {
+      this.setState({
+        fileList: [file],
+        uploadStatus: STATUS.READY,
+      });
+    } else {
+      this.setState(defaultState);
+      message.error(`${file.name} has an invalid extension`);
+    }
+
+    // false stops them from automatically using their upload functionaility
+    return false;
+  };
+
+  startUpload = async () => {
+    const { assetId } = this.props;
+    if (this.state.uploadStatus !== STATUS.READY) {
+      return;
+    }
+
+    try {
+      await this.props.beforeUploadStart();
+    } catch (e) {
+      this.props.onUploadFailure('Unable to start upload');
+      return;
+    }
+
+    message.info('Starting Upload...');
+
+    const file = this.state.fileList[0];
+
+    const mimeType = this.getMIMEType(file.name);
+    if (!mimeType) {
+      this.props.onUploadFailure('Unable to detect file type');
+      return;
+    }
+
+    const fileMetadata = (await sdk.files.upload({
+      name: file.name,
+      mimeType,
+      source: 'Discovery',
+      ...(assetId && { assetIds: [assetId] }),
+    })) as UploadFileMetadataResponse;
+    const { uploadUrl, id } = fileMetadata;
+
+    if (!uploadUrl || !id) {
+      this.props.onUploadFailure('Unable to create file');
+      return;
+    }
+
+    const chunkMultiple = Math.min(
+      Math.max(
+        2, // 0.5MB min chunks
+        Math.ceil((file.size / 20) * 262144) // will divide into 20 segments
+      ),
+      200 // 50 MB max
+    );
+
+    this.currentUpload = new UploadGCS({
+      id: 'discovery-upload',
+      url: uploadUrl,
+      file,
+      chunkSize: 262144 * chunkMultiple,
```

Comment:
wat

Reply:
```
{
  id: null, // required - a unique ID for the upload 
  url: null, // required - GCS resumable URL
  file: null, // required - instance of File
  chunkSize: 262144, // optional - chunk size must be a multiple of 262144
  storage: window.localStorage, // optional - storage mechanism used to persist chunk meta data
  contentType: 'text/plain', // optional - content type of the file being uploaded
  onChunkUpload: () => {} // optional - a function that will be called with progress information
}
```
from https://github.com/QubitProducts/gcs-browser-upload

Reply:
I mean the magic number. Got it. Can you add link to that (maybe you did)